src/program.rs

```diff
@@ -401,70 +405,178 @@ impl From<ColorOpt> for ColorChoice {
         let term = cache.get_ref(file_id).unwrap();
-        let document = AstNode::from(NodeValue::Document);
 
-        // Our nodes in the Markdown document are owned by this arena
-        let arena = Arena::new();
+        let Some(docs) = extract_documentation(term) else {
+            return Ok(());
+        };
 
-        // The default ComrakOptions disables all extensions (essentially reducing to CommonMark)
-        let options = ComrakOptions::default();
+        match format {
+            DocFormat::Markdown => {
+                let document = AstNode::from(NodeValue::Document);
 
-        to_markdown(term, 0, &arena, &document, &options)?;
-        format_commonmark(&document, &options, out)
-            .map_err(|e| Error::IOError(IOError(e.to_string())))?;
+                // Our nodes in the Markdown document are owned by this arena
+                let arena = Arena::new();
 
-        Ok(())
+                // The default ComrakOptions disables all extensions (essentially reducing to CommonMark)
+                let options = ComrakOptions::default();
+
+                to_markdown(&docs, 0, &arena, &document, &options)?;
+                format_commonmark(&document, &options, out)
+                    .map_err(|e| Error::IOError(IOError(e.to_string())))?;
+
+                Ok(())
+            }
+            DocFormat::Json => serde_json::to_writer(out, &docs)
+                .map_err(|e| Error::SerializationError(SerializationError::Other(e.to_string()))),
+        }
     }
 
-    /// Recursively walk the given richterm, recursing into fields of record, looking for documentation.
+    /// Recursively walk the given `DocOutput`, recursing into fields, looking for documentation.
     /// This documentation is then added to the provided document.
     fn to_markdown<'a>(
-        rt: &'a RichTerm,
+        docs: &'a ExtractedDocumentation,
         header_level: u8,
         arena: &'a Arena<AstNode<'a>>,
         document: &'a AstNode<'a>,
         options: &ComrakOptions,
     ) -> Result<(), Error> {
-        match rt.term.as_ref() {
-            Term::Record(record) | Term::RecRecord(record, _, _) => {
-                // Sorting fields for a deterministic output
-                let mut entries: Vec<(_, _)> = record.fields.iter().collect();
-                entries.sort_by_key(|(k, _)| *k);
-
-                for (ident, field) in entries {
-                    let header = mk_header(ident.label(), header_level + 1, arena);
-                    document.append(header);
-
-                    if let Some(ref doc) = field.metadata.doc {
-                        document.append(parse_documentation(header_level, arena, doc, options));
-                    }
-
-                    if let Some(ref value) = field.value {
-                        to_markdown(value, header_level + 1, arena, document, options)?;
-                    }
-                }
+        let mut entries: Vec<(_, _)> = docs.fields.iter().collect();
```

Comment:
It's a matter of taste (so just a comment, not an actual review suggestion), but I found the original version with newlines to be more readable.